certbot/compat/filesystem.py

```diff
@@ -36,6 +41,43 @@ def chmod(file_path, mode):
         _apply_win_mode(file_path, mode)
 
 
+def open(file_path, flags, mode=0o777):  # pylint: disable=redefined-builtin
+    # type: (str, int, int) -> int
+    """
+    Wrapper of original os.open function, that will ensure on Windows that given mode
+    is correctly applied.
+    :param str file_path: The file path to open
+    :param int flags: Flags to apply on file while opened
+    :param int mode: POSIX mode to apply on file when opened,
+        Python defaults will be applied if ``None``
+    :returns: the file descriptor to the opened file
+    :rtype: int
+    """
+    if not POSIX_MODE:
+        # Handle creation of the file atomically with proper permissions
+        if not os.path.exists(file_path) and flags & os.O_CREAT:
+            attributes = win32security.SECURITY_ATTRIBUTES()
+            security = attributes.SECURITY_DESCRIPTOR
+            user = _get_current_user()
+            dacl = _generate_dacl(user, mode)
+            security.SetSecurityDescriptorDacl(1, dacl, 0)
+
+            handle = win32file.CreateFile(file_path, win32file.GENERIC_READ, 0,
+                                          attributes, win32con.CREATE_NEW, 0, None)
+            handle.Close()
+
+            # At this point, the file that was not existing has been created, with proper
+            # permissions, so os.O_CREAT and os.O_EXCL are not needed anymore. We remove them
+            # from the flags to avoid a FileExists exception when os.open will be called.
+            flags = flags ^ os.O_CREAT ^ os.O_EXCL
+        elif os.path.exists(file_path):
```

Comment:
where did all the `os.O_EXCL` handling go? if the file already exists but `os.O_EXCL` was set, won't we chmod then open the file anyway?

Reply:
Indeed, see my general comment below.